cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MainOutputCommitter.java

```diff
@@ -217,24 +217,18 @@ private void onFinish(JobContext jobContext, boolean success) throws IOException
    * Either calls onSuccess or onFailure on all of the DatasetOutputCommitters.
    */
   private void finishDatasets(final JobContext jobContext, final boolean success) throws Exception {
-    taskContext.execute(new TxRunnable() {
-      @Override
-      public void run(DatasetContext ctxt) throws Exception {
-        ClassLoader oldClassLoader =
-          ClassLoaders.setContextClassLoader(jobContext.getConfiguration().getClassLoader());
-        Map<String, DatasetOutputCommitter> datasetOutputCommitters = getDatasetOutputCommitters(outputs);
-        try {
-          if (success) {
-            commitOutputs(datasetOutputCommitters);
-          } else {
-            // but this output committer failed: call onFailure() for all committers
-            failOutputs(datasetOutputCommitters);
-          }
-        } finally {
-          ClassLoaders.setContextClassLoader(oldClassLoader);
-        }
+    ClassLoader oldClassLoader = ClassLoaders.setContextClassLoader(jobContext.getConfiguration().getClassLoader());
+    Map<String, DatasetOutputCommitter> datasetOutputCommitters = getDatasetOutputCommitters(outputs);
+    try {
+      if (success) {
+        commitOutputs(datasetOutputCommitters);
+      } else {
+        // but this output committer failed: call onFailure() for all committers
+        failOutputs(datasetOutputCommitters);
       }
-    });
+    } finally {
+      ClassLoaders.setContextClassLoader(oldClassLoader);
+    }
```

Comment:
Pushed a commit where I also perform operation rollback in the `onFailure` method of the PartitionedFileSet. Does that look good to you?